### PR TITLE
Add named __construct() params

### PR DESCRIPTION
**Pull request by yiiliveext**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Tests pass?   | ✔️

### CODE REVIEW

File: src/Definitions/ArrayBuilder.php [hunk 1]
```diff
@@ -5,6 +5,7 @@
 use Psr\Container\ContainerInterface;
 use Yiisoft\Factory\Exceptions\NotInstantiableException;
 use Yiisoft\Factory\Resolvers\ClassNameResolver;
+use Yiisoft\Factory\Tests\Support\EngineInterface;
```
Comment (samdark) [inline src/Definitions/ArrayBuilder.php:8]: Should not be imported outside of tests.

File: src/Definitions/ArrayBuilder.php [hunk 2]
```diff
@@ -19,15 +20,26 @@ public function build(ContainerInterface $container, ArrayDefinition $definition
...
         if (!empty($definition->getParams())) {
-            foreach (array_values($definition->getParams()) as $index => $param) {
-                if ($param instanceof DefinitionInterface) {
-                    $dependencies[$index] = $param;
+            foreach ($definition->getParams() as $index => $param) {
+                if ($param instanceof ReferenceInterface) {
+                    if (is_string($index)) {
+                        $dependencies[$index] = $param;
+                    } else {
+                        $dependencies[key($dependencies)] = $param;
```
Comment (samdark) [inline src/Definitions/ArrayBuilder.php:28]: It is better to make it behave the same way as in Injector:

1. Named ones are set by name.
2. Typed ones are set by type (if type is class/interface).
3. Othrewise, ignore parameter.

File: bash.exe.stackdump [hunk 3]
```diff
@@ -0,0 +1,4 @@
+Stack trace:
+Frame     Function  Args
```
[the remark this answers is not in the document]
Comment (yiiliveext) [reply bash.exe.stackdump]: Removed

**Comment by xepozz** [comment]
Tell me please, which problem will you solve by this PR?

**Comment by yiiliveext** [comment]
> Tell me please, which problem will you solve by this PR?

1. Named params.
2. Params indexed by a position

**Pull request merged**